test/src/unit-capi-smoke-test.cc

```diff
@@ -0,0 +1,757 @@
+    tiledb_array_type_t array_type,
+    tiledb_layout_t cell_order,
+    tiledb_layout_t tile_order,
+    tiledb_layout_t write_order,
+    tiledb_encryption_type_t encryption_type) {
+  const string array_name = "dynamic_array";
+
+  // Skip row-major and col-major writes for sparse arrays.
+  if (array_type == TILEDB_SPARSE &&
+      (write_order == TILEDB_ROW_MAJOR || write_order == TILEDB_COL_MAJOR)) {
+    return;
+  }
+
+  // String_ascii, float32, and float64 types can only be
+  // written to sparse arrays.
+  if (array_type == TILEDB_DENSE && (test_attr.type_ == TILEDB_STRING_ASCII ||
+                                     test_attr.type_ == TILEDB_FLOAT32 ||
+                                     test_attr.type_ == TILEDB_FLOAT64)) {
+    return;
+  }
+
+  // Create the array.
+  create_array(
+      array_name,
+      array_type,
+      test_dims,
+      test_attr,
+      cell_order,
+      tile_order,
+      encryption_type);
+
+  // Define the write query buffers for "a".
+  vector<test_query_buffer_t> write_query_buffers;
+  uint64_t total_cells = 1;
+  for (const auto& test_dim : test_dims) {
+    const uint64_t max_range = ((uint64_t*)(test_dim.domain_))[1];
+    const uint64_t min_range = ((uint64_t*)(test_dim.domain_))[0];
+    const uint64_t full_range = max_range - min_range + 1;
+    total_cells *= full_range;
+  }
+
+  uint64_t a_write_buffer_size;
+  int32_t* a_write_buffer = nullptr;
+
+  if (test_attr.cell_val_num_ == TILEDB_VAR_NUM) {
+    a_write_buffer_size =
+        total_cells * 2 * tiledb_datatype_size(test_attr.type_);
+    a_write_buffer = (int32_t*)tdb_malloc(a_write_buffer_size);
+    for (uint64_t i = 0; i < (total_cells * 2); i++) {
+      a_write_buffer[i] = i;
+    }
+  } else {
+    a_write_buffer_size = total_cells * tiledb_datatype_size(test_attr.type_);
+    a_write_buffer = (int32_t*)tdb_malloc(a_write_buffer_size);
+    for (uint64_t i = 0; i < total_cells; i++) {
+      a_write_buffer[i] = i;
+    }
+  }
+
+  uint64_t a_write_buffer_offset_size = total_cells * sizeof(uint64_t);
+  uint64_t* a_write_buffer_offset =
+      (uint64_t*)tdb_malloc(a_write_buffer_offset_size);
+  for (uint64_t i = 0; i < total_cells; i++) {
+    a_write_buffer_offset[i] = i * tiledb_datatype_size(test_attr.type_) * 2;
+  }
+
+  // Variable buffer is set only when the attribute is var-sized.
+  if (test_attr.cell_val_num_ == TILEDB_VAR_NUM) {
+    write_query_buffers.emplace_back(
+        test_attr.name_,
+        a_write_buffer_offset,
+        &a_write_buffer_offset_size,
+        a_write_buffer,
+        &a_write_buffer_size);
+  } else {
+    write_query_buffers.emplace_back(
+        test_attr.name_,
+        a_write_buffer,
+        &a_write_buffer_size,
+        nullptr,
+        nullptr);
+  }
+
+  // Define dimension query write vectors for either sparse arrays
+  // or dense arrays with an unordered write order.
+  vector<uint64_t*> d_write_buffers;
+  uint64_t write_buffer_size = 0;
```

Comment:
Maybe I'm thinking about the logic incorrectly here. It could be more beneficial to just define the `write_buffer_size` above the following code block.

```
// Fill the write buffers with values.
    auto dims_iter = test_dims.begin();
    for (uint64_t* const d_write_buffer : d_write_buffers) {
      for (uint64_t i = 0; i < dimension_ranges[i]; i++) {
        d_write_buffer[i] = i;
      }
      write_buffer_size = 
        dim_buffer_len * tiledb_datatype_size(dims_iter->type_);

      write_query_buffers.emplace_back(
          dims_iter->name_,
          d_write_buffer,
          &(write_buffer_size),
          nullptr,
          nullptr);
      dims_iter = std::next(dims_iter, 1);
    }
  }
  ```
  My idea here being that the `write_buffer_size` is calculated solely for each `d_write_buffer` to be placed into the `write_query_buffers`. Is this sufficient? 
  
  I can see what you're saying about the size getting rewritten. Probably better to just say here `uint64_t write_buffer_size = dim_buffer_len * tiledb_datatype_size(dims_iter->type_);` inside this for loop and skip defining it above.